src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.Streaming.cs

```diff
@@ -99,11 +98,35 @@ private void SendBatchAsStreamingUpdate(in RenderBatch renderBatch, TextWriter w
             var componentIdsInDepthOrder = bufSizeRequired < 1024
                 ? MemoryMarshal.Cast<byte, ComponentIdAndDepth>(stackalloc byte[bufSizeRequired])
                 : new ComponentIdAndDepth[count];
+            var includedComponentCount = 0;
             for (var i = 0; i < count; i++)
             {
                 var componentId = renderBatch.UpdatedComponents.Array[i].ComponentId;
-                componentIdsInDepthOrder[i] = new(componentId, GetComponentDepth(componentId));
+                var componentState = (EndpointComponentState)GetComponentState(componentId);
+
+                if (componentState.IsInteractive && componentState.WasIncludedInStreamingResponse)
+                {
+                    // If a component with an interactive render mode was already included in a
+                    // streaming response, then don't include it again. This minimizes the possibility
+                    // that a stream rendering update forces an interactive component to be re-initialized
+                    // after interactivity has already begun.
+
+                    // It should be noted that a non-interactive parent component might still SSR the
+                    // interactive component again, at which point the browser will remove and
+                    // re-initialize the component with updated parameter values.
+                    continue;
```

Comment:
I've been thinking through whether or not we should prerender interactive components more than once.

At first I thought it was debatable and could go either way, but eventually realised that **we don't want components to start becoming interactive until streaming SSR is completed**. Because otherwise, we'd be wastefully discarding any "loading" work that is happening during prerendering, and would be unable to initialize the interactive component with any `PersistentComponentState` that is obtained from loading.

If that's true, then I don't see any reason *not* to send streaming updates to the prerendered content for all components equally, whether or not they are interactive. The server is doing this work to load and render data for them anyway; presumably the app developer doesn't want to discard the output from that work.

The more subtle case is if we know this request is for a progressively-enhanced navigation (so there may already be interactive components on the page). We do have a special HTTP header telling us if it's a PE-enhanced nav, and in an ideal situation, the server would somehow know which rendermode boundaries can now be ignored (because there's already an interactive version of that component in the browser) and would just emit updated parameters for them. But unfortunately the server *doesn't* know which rendermode boundaries can be ignored - maybe we're rendering a different set of components now. The only thing that can resolve this correspondence is the client, which will use its DOM sync logic to establish whether the newly-received interactive markers line up with existing interactive components, and in that case, it should just take the new parameters and discard the prerendered content. Clearly there's some wastefulness in that but AFAIK it's the best that's logically possible, or maybe we think of further optimizations in the future whereby the PE-nav requests somehow also communicate info about what interactive components they already have (not for .NET 8).

What do you think? Does this sound right or am I missing something?

Reply:
Thanks for the feedback, @SteveSandersonMS!

I think I agree with you that newly-SSR'd components shouldn't become active until streaming rendering completes. I also went back and forth on this, and my initial reasoning for eagerly-initializing interactive components was:
* It might inhibit the user experience to delay interactivity for components that might not receive further stream rendering updates anyway.
* If a fully initialized version of a component looks significantly different than its initial render, making the component interactive as soon as possible minimizes how much the component reverts back to its "initial" state when interactivity finally starts.

However, the first point is probably not a big deal because unless stream rendering takes ages, the user experience isn't likely to degrade massively by a component not being interactive for a few seconds. And the second point is probably not a concern considering the point you made about `PersistentComponentState`.

---

I also agree with your analysis about what would be the most optimal behavior regarding the interaction between SSR and interactive components. Part of my reasoning for the current approach was that it made the logic super straightforward when an SSR update conflicts with an interactive component (SSR always wins, dispose and re-initialize the interactive component). I had prototyped some approaches for a more sophisticated merging behavior where the DOM sync algorithm would correctly know to ignore existing interactive parts of the DOM, which it would correctly map to incoming component comments, but I wasn't happy with any of those approaches.

I did some more thinking about this today and started implementing an approach that I believe should work fairly well. Most of the ideas in my current approach still apply. It’s not ready to commit yet but here's a summary of what I'm thinking:

* If a component gets SSR'd, it always produces output in the response, regardless of whether it was rendered already.
* Before the DOM sync occurs (in the `beforeDocumentUpdated` callback), we:
  * Find component markers in the "incoming" HTML using existing logic (`discoverComponents`).
  * Find interactive component trees in the "destination" content being updated, and move them out of the DOM to separate document fragments
    * This is similar to what happens in the current implementation, minus the eager component disposal
    * This allows the DOM sync algorithm to work without special logic handling interactively managed DOM
    * _Note:_ We won't have to handle SSR updates that involve only a subtree of interactive DOM. When a component becomes interactive, prerendered content (including boundary markers) gets removed, and the current stream rendering logic already ignores incoming content that doesn't have matching boundary markers.
    * _Another note:_ I don't anticipate that removing/reinserting chunks of DOM like this will create any visual flickering or shifting because all changes to the DOM here happen synchronously.
  * For each interactive root component previously identified in the destination:
    * If it has a corresponding component comment in the incoming markup, clear any associated prerendered content (leaving the comment intact).
    * Otherwise, initiate asynchronous disposal of the component.
* After the DOM sync occurs (in the `afterDocumentUpdated` callback), we:
  * Move existing interactive components back into the DOM at the locations of their corresponding newly inserted component comments
    * Apply updated parameters from the comment's payload.
    * Remove the comment.
  * Initialize interactive components for each comment that didn't correlate to an existing interactive component.

---

To make this work, we'll probably have to serialize a `componentId` in each component comment, representing the ID of the SSR'd component. That way, interactive components can be mapped to SSR'd comments.

Also, the Blazor Server component descriptor protocol currently has a security restriction that forces root components to be added in the order that they get rendered. This isn't always possible with SSR because:
* We might deliberately try to utilize SSR'd components in a different order than they were rendered (for example, using SSR output to update component parameters might happen before newly SSR'd components become interactive).
* SSR'd components might get ignored for interactivity outright (for example, a component gets added and then removed from the page in two separate SSR updates before interactivity starts).

We should be able to get around this problem by allowing Blazor Server component comments to be used in the "wrong" order, but then requiring the client to provide the sequence number of each component along with the data-protected payload. From a security perspective, I think this is equivalent to what we have today. The only difference is that instead of the sequence numbers being implied by the order of each component descriptor in the serialized array, they're stated explicitly.

**TL;DR:** I agree with your comment and I'm currently working on updates 🙂